onnxruntime/contrib_ops/cuda/decoder/decoder_masked_multihead_attention.cc

```diff
@@ -0,0 +1,187 @@
+using namespace onnxruntime::cuda;
+using namespace ::onnxruntime::common;
+using namespace ONNX_NAMESPACE;
+
+namespace onnxruntime {
+namespace contrib {
+namespace cuda {
+
+constexpr int kPastSequenceLengthInputIndex = 6;
+constexpr int kPastInputIndex = 4;
+constexpr int kPresentOutputIndex = 1;
+
+#define REGISTER_KERNEL_TYPED(T1, T2)                                          \
+  ONNX_OPERATOR_TYPED_KERNEL_EX(                                               \
+      DecoderMaskedMultiheadAttention,                                         \
+      kMSDomain,                                                               \
+      1,                                                                       \
+      T1,                                                                      \
+      kCudaExecutionProvider,                                                  \
+      (*KernelDefBuilder::Create())                                            \
+          .MayInplace(kPastInputIndex, kPresentOutputIndex)                    \
+          .TypeConstraint("T", DataTypeImpl::GetTensorType<T1>())              \
+          .InputMemoryType(OrtMemTypeCPUInput, kPastSequenceLengthInputIndex), \
+      DecoderMaskedMultiheadAttention<T1, T2>);
+
+REGISTER_KERNEL_TYPED(float, float)
+REGISTER_KERNEL_TYPED(MLFloat16, uint16_t)
+
+template <typename T1, typename T2>
+Status DecoderMaskedMultiheadAttention<T1, T2>::ComputeInternal(OpKernelContext* context) const {
+  const Tensor* input = context->Input<Tensor>(0);
+  const Tensor* weights = context->Input<Tensor>(1);
+  const Tensor* bias = context->Input<Tensor>(2);
+  const Tensor* mask_index = context->Input<Tensor>(3);
+  const Tensor* past = context->Input<Tensor>(kPastInputIndex);
+  const Tensor* relative_position_bias = context->Input<Tensor>(5);
+  const Tensor* past_seq_len = context->Input<Tensor>(kPastSequenceLengthInputIndex);
+
+  auto& device_prop = GetDeviceProp();
+  DecoderMaskedMultiheadAttentionParams parameters;
+  ORT_RETURN_IF_ERROR(CheckInputs(input->Shape(),
+                                  weights->Shape(),
+                                  bias->Shape(),
+                                  mask_index,
+                                  past,
+                                  relative_position_bias,
+                                  &parameters,
+                                  device_prop.maxThreadsPerBlock,
+                                  past_seq_len));
+
+  // Sanity check
+  ORT_ENFORCE(past_present_share_buffer_);
+
+  int batch_size = parameters.batch_size;
+  int sequence_length = parameters.sequence_length;
+
+  // This kernel is for decoding only (i.e.) sequence length has to be 1
+  if (sequence_length != 1) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Input sequence length should be 1 to use DecoderMaskedMultiheadAttention");
+  }
+
+  // TODO(hasesh): In future, we may support CrossAttention. Currently, this kernel only supports SelfAttention.
```

Comment:
There are some TODOs here that can be enhanced later based on model requirements - Currently, these are unsupported.